### PR TITLE
Optimize thread pool

### PR DESCRIPTION
**Pull request by jacquesqiao**
refs: 
https://stackoverflow.com/questions/44387584/double-free-or-corruption-prev-error-when-working-with-threads-and-mutex
https://stackoverflow.com/questions/17101922/do-i-have-to-acquire-lock-before-calling-condition-variable-notify-one
https://www.justsoftwaresolutions.co.uk/threading/implementing-a-thread-safe-queue-using-condition-variables.html

### CODE REVIEW

**Comment by typhoonzero** [comment]
May be use thread pool under thirdparty is OK.

**Review by typhoonzero — APPROVED** [review]
Maybe move on to thirdparty implementation later.

**Comment by panyx0718** [comment]
what's the gain of this optimization?

File: paddle/fluid/framework/threadpool.h [hunk 1]
```diff
@@ -84,7 +83,13 @@ class ThreadPool {
       return nullptr;
     });
     std::future<std::unique_ptr<platform::EnforceNotMet>> f = task.get_future();
-    tasks_.push(std::move(task));
+    {
+      std::unique_lock<std::mutex> lock(mutex_);
+      if (!running_) {
```
Comment (panyx0718) [inline paddle/fluid/framework/threadpool.h:88]: Is this `running_` check needed?

File: paddle/fluid/framework/threadpool.cc [hunk 2]
```diff
@@ -57,10 +57,10 @@ ThreadPool::ThreadPool(int num_threads) : running_(true) {
 ThreadPool::~ThreadPool() {
   {
     // notify all threads to stop running
-    std::lock_guard<std::mutex> l(mutex_);
+    std::unique_lock<std::mutex> l(mutex_);
```
Comment (panyx0718) [inline paddle/fluid/framework/threadpool.cc:60]: is unique_lock faster?

File: paddle/fluid/framework/threadpool.cc [hunk 3]
```diff
@@ -57,10 +57,10 @@ ThreadPool::ThreadPool(int num_threads) : running_(true) {
...
     running_ = false;
-    scheduled_.notify_all();
   }
+  scheduled_.notify_all();
```
Comment (panyx0718) [inline paddle/fluid/framework/threadpool.cc:63]: some doc said notify_all doesn't need lock. But in reality, I see lots of projects locking this

File: paddle/fluid/framework/threadpool.cc (hunk 2 above)
Comment (jacquesqiao) [reply paddle/fluid/framework/threadpool.cc:60]: No, here just want to be consistent with other place.

File: paddle/fluid/framework/threadpool.h (hunk 1 above)
Comment (jacquesqiao) [reply paddle/fluid/framework/threadpool.h:88]: I think we need it, but it can work if not add.

**Pull request merged**